tests/functional_tests/mining.py

```diff
@@ -97,17 +107,59 @@ def mine(self, via_daemon):
         # wait till we mined a few of them
         target_height = prev_height + 5
         height = prev_height
-        timeout = 240 # randomx is slow to init
+        
+        """
+        Randomx init has high variance on CI machines due to noisy neighbors, 
+        taking up resources in parallel (including by our own jobs).
+        
+        Mining is organized in the following scheme:
+            1) first loop's pass: RandomX init and mining
+            2) every next pass:   only mining
+        Pass 1) takes much more time than pass 2)
+        Pass 1) uses all cores, pass 2) just one (currently)
+        For the above reasons both passes need separate timeouts and adjustments.
+        After the first pass, the timeout is being reset to a lower value.
+        """
+        
+        def calc_timeout(seconds_constant, time_pi, cores):
+            """
+            The time it took to calculate pi under certain conditions 
+            is proportional to the time it will take to calculate the real job.
+            
+            The number of cores used decreases the time almost linearly.
+            """
+            timeout = float(seconds_constant) * time_pi / float(cores)
+            return timeout
+
+        timeout_base_init = 60 # RX init needs more time
+        timeout_base_mine = 20
+        cores_init = multiprocessing.cpu_count() # RX init uses all cores
+        cores_mine = res_status.threads_count    # Mining uses parametric number of cores
+        timeout_init = calc_timeout(timeout_base_init, time_pi_all_cores,  cores_init)
+        timeout_mine = calc_timeout(timeout_base_mine, time_pi_single_cpu, cores_mine)
+        
+        msg = "Timeout for {} adjusted for the currently available CPU power, is {} s"
+        print(msg.format("init,  ", round(timeout_init, 1)))
+        print(msg.format("mining,", round(timeout_mine, 1)))
+        
+        timeout = timeout_init
+        init_rx = False
         while height < target_height:
             seen_height = height
-            for _ in range(timeout):
+            for _ in range(int(math.ceil(timeout))):
                 time.sleep(1)
+                seconds_passed = monotonic.monotonic() - start
                 height = daemon.get_info().height
                 if height > seen_height:
                     break
             else:
-                assert False, 'Failed to mine successor to block %d (initial block = %d)' % (seen_height, prev_height)
-            timeout = 10
+                assert False, 'Failed to mine successor to block %d (initial block = %d) after %d s. RX initialized = %r' % (seen_height, prev_height, round(seconds_passed), init_rx)
+            if not init_rx:
```

Comment:
You don't really need this boolean, can just use `seen_height == prev_height`.

Reply:
I need it for readability. Same as with documentation and self-documenting code: the goal is not only to write the most minimal code, but code that can be undrestood by others.

Reply:
I don't find the variable terribly offensive or anything, but it seems you're perhaps papering over the real readability issue, which is, I think, that `prev_height` is a rather poor name, and should be something along the lines or `initial_height`. With that change, I think the conditional proves quite clear without an extra variable, but I won't belabor the point.

Reply:
The `initial_height` is indeed a much better name for the variable, especially since it's a constant.

Still the `rx_inited` var is used twice - once in the print function and secondly below the `for` loop. Using the comparison twice is error-prone IMHO.

BTW. I added an additional printout, which will be executed if one `while` loop's iteration is not enough.